File: cpp/FMU-proxy/src/fmuproxy/grpc/client/grpc_fmu_client.cpp
/*
 * The MIT License
 *
 * Copyright 2017-2018 Norwegian University of Technology
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING  FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */


#include "../../util/file_util.hpp"
#include "grpc_client_helper.hpp"

#include <fmuproxy/grpc/client/grpc_fmu_client.hpp>

#include <boost/filesystem.hpp>
#include <utility>

using namespace grpc;
using namespace fmuproxy::grpc;
using namespace fmuproxy::grpc::client;

namespace fs = boost::filesystem;

grpc_fmu_client::grpc_fmu_client(const std::string& host, const unsigned int port)
{
    const auto channel = CreateChannel(host + ":" + to_string(port), InsecureChannelCredentials());
    stub_ = move(FmuService::NewStub(channel));
}

remote_grpc_fmu grpc_fmu_client::from_url(const std::string& url)
{
    ClientContext ctx;
    Url url_;
    FmuId fmuId;
    url_.set_url(url);
    stub_->LoadFromUrl(&ctx, url_, &fmuId);
    return from_guid(fmuId.value());
}

remote_grpc_fmu grpc_fmu_client::from_file(const std::string& file)
{
    fs::path p(file);
    std::string name = p.stem().string();

    std::string data;
    read_data(file, data);

    ClientContext ctx;
    File file_;
    FmuId fmuId;
    file_.set_name(name);
    file_.set_data(data);

    stub_->LoadFromFile(&ctx, file_, &fmuId);
    return from_guid(fmuId.value());
}

remote_grpc_fmu grpc_fmu_client::from_guid(const std::string& guid)
{
    return remote_grpc_fmu(guid, stub_);
}

remote_grpc_fmu::remote_grpc_fmu(string  fmuId, std::shared_ptr<FmuService::Stub> stub)
    : fmuId_(std::move(fmuId))
    , stub_(std::move(stub))
{}

std::shared_ptr<const fmi4cpp::fmi2::cs_model_description>& remote_grpc_fmu::getModelDescription()
{
    if (modelDescription_ == nullptr) {
        ClientContext ctx;
        GetModelDescriptionRequest request;
        request.set_fmu_id(fmuId_);
        fmuproxy::grpc::ModelDescription md;
        stub_->GetModelDescription(&ctx, request, &md);
        modelDescription_ = std::move(convert(md));
    }
    return modelDescription_;
}

std::unique_ptr<remote_fmu_slave> remote_grpc_fmu::newInstance()
{
    ClientContext ctx;
    InstanceId instance_id;
    CreateInstanceRequest request;
    request.set_fmu_id(fmuId_);
    stub_->CreateInstance(&ctx, request, &instance_id);
    return std::make_unique<remote_fmu_slave>(instance_id.value(), *stub_, getModelDescription());
}

File: fmu-proxy-clients/src/main/kotlin/no/ntnu/ihb/fmuproxy/grpc/GrpcFmuClient.kt
/*
 * The MIT License
 *
 * Copyright 2017-2018 Norwegian University of Technology (NTNU)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING  FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package no.ntnu.ihb.fmuproxy.grpc

import com.google.protobuf.ByteString
import io.grpc.ManagedChannelBuilder
import no.ntnu.ihb.fmi4j.*
import no.ntnu.ihb.fmi4j.modeldescription.*
import no.ntnu.ihb.fmuproxy.AbstractRpcFmuClient
import no.ntnu.ihb.fmuproxy.DirectionalDerivativeResult
import no.ntnu.ihb.fmuproxy.InstanceId
import no.ntnu.ihb.fmuproxy.StepResult
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.Closeable
import java.io.File
import java.io.FileInputStream
import java.net.URL
import java.util.concurrent.TimeUnit

/**
 * @author Lars Ivar Hatledal
 */
class GrpcFmuClient(
        host: String,
        port: Int
) : Closeable {

    private val channel = ManagedChannelBuilder
            .forAddress(host, port)
            .usePlaintext()
            .build()

    private val stub = FmuServiceGrpc.newBlockingStub(channel)

    fun load(fmuId: String): AbstractRpcFmuClient {
        return GrpcFmu(fmuId);
    }

    fun load(url: URL): AbstractRpcFmuClient {
        return Service.Url.newBuilder()
                .setUrl(url.toString())
                .build().let {
                    load(stub.loadFromUrl(it).value)
                }
    }

    fun load(file: File): AbstractRpcFmuClient {
        if (!file.name.endsWith(".fmu")) {
            throw IllegalArgumentException("File must be an FMU!")
        }
        val data = FileInputStream(file).use {
            ByteString.copyFrom(it.readBytes())
        }
        return Service.File.newBuilder()
                .setName(file.nameWithoutExtension)
                .setData(data)
                .build().let {
                    load(stub.loadFromFile(it).value)
                }
    }

    override fun close() {
        channel.shutdownNow()
        channel.awaitTermination(500, TimeUnit.MILLISECONDS)
    }

    private inner class GrpcFmu(
            fmuId: String
    ) : AbstractRpcFmuClient(fmuId) {

        override val implementationName: String = "GrpcClient"

        override val modelDescription: CoSimulationModelDescription by lazy {
            Service.GetModelDescriptionRequest.newBuilder()
                    .setFmuId(fmuId)
                    .build().let {
                        stub.getModelDescription(it).convert()
                    }
        }

        override fun createInstance(): String {
            return Service.CreateInstanceRequest.newBuilder()
                    .setFmuId(fmuId)
                    .build().let {
                        stub.createInstance(it).value
                    }
        }

        override fun setupExperiment(instanceName: InstanceId, start: Double, stop: Double, tolerance: Double): FmiStatus {
            return Service.SetupExperimentRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .setStart(start)
                    .setStop(stop)
                    .setTolerance(tolerance)
                    .build().let {
                        stub.setupExperiment(it).convert()
                    }
        }

        override fun enterInitializationMode(instanceName: InstanceId): FmiStatus {
            return Service.EnterInitializationModeRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .build().let {
                        stub.enterInitializationMode(it).convert()
                    }
        }

        override fun exitInitializationMode(instanceName: InstanceId): FmiStatus {
            return Service.ExitInitializationModeRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .build().let {
                        stub.exitInitializationMode(it).convert()
                    }
        }

        override fun step(instanceName: String, stepSize: Double): StepResult {
            return Service.StepRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .setStepSize(stepSize)
                    .build().let { req ->
                        stub.step(req).let { response ->
                            StepResult(response.simulationTime, response.status.convert())
                        }
                    }
        }

        override fun reset(instanceName: String): FmiStatus {
            return Service.ResetRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .build().let {
                        stub.reset(it).status.convert()
                    }
        }

        override fun terminate(instanceName: String): FmiStatus {
            return Service.TerminateRequest.newBuilder().setInstanceId(instanceName).build().let {
                stub.terminate(it).status.convert()
            }
        }

        override fun freeInstance(instanceName: InstanceId) {
            return Service.FreeRequest.newBuilder().setInstanceId(instanceName).build().let {
                stub.freeInstance(it)
            }
        }

        override fun readInteger(instanceName: String, vr: List<ValueReference>): IntegerArrayRead {
            return stub.readInteger(getReadRequest(instanceName, vr)).convert()
        }

        override fun readReal(instanceName: String, vr: List<ValueReference>): RealArrayRead {
            return stub.readReal(getReadRequest(instanceName, vr)).convert()
        }

        override fun readString(instanceName: String, vr: List<ValueReference>): StringArrayRead {
            return stub.readString(getReadRequest(instanceName, vr)).convert()
        }

        override fun readBoolean(instanceName: String, vr: List<ValueReference>): BooleanArrayRead {
            return stub.readBoolean(getReadRequest(instanceName, vr)).convert()
        }

        override fun writeInteger(instanceName: String, vr: List<ValueReference>, value: List<Int>): FmiStatus {
            return Service.WriteIntegerRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .addAllValueReferences(vr)
                    .addAllValues(value)
                    .build().let { request ->
                        stub.writeInteger(request).convert()
                    }
        }


        override fun writeReal(instanceName: String, vr: List<ValueReference>, value: List<Real>): FmiStatus {
            return Service.WriteRealRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .addAllValueReferences(vr)
                    .addAllValues(value)
                    .build().let { request ->
                        stub.writeReal(request).convert()
                    }
        }

        override fun writeString(instanceName: String, vr: List<ValueReference>, value: List<String>): FmiStatus {
            return Service.WriteStringRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .addAllValueReferences(vr)
                    .addAllValues(value)
                    .build().let { request ->
                        stub.writeString(request).convert()
                    }
        }

        override fun writeBoolean(instanceName: String, vr: List<ValueReference>, value: List<Boolean>): FmiStatus {
            return Service.WriteBooleanRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .addAllValueReferences(vr)
                    .addAllValues(value)
                    .build().let { request ->
                        stub.writeBoolean(request).convert()
                    }
        }

        override fun getDirectionalDerivative(instanceName: InstanceId, vUnknownRef: List<ValueReference>, vKnownRef: List<ValueReference>, dvKnownRef: List<Double>): DirectionalDerivativeResult {
            return Service.GetDirectionalDerivativeRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .addAllVUnknownRef(vUnknownRef)
                    .addAllVKnownRef(vKnownRef)
                    .addAllDvKnownRef(dvKnownRef).build().let { request ->
                        stub.getDirectionalDerivative(request).let { response ->
                            DirectionalDerivativeResult(response.dvUnknownRefList.toDoubleArray(), response.status.convert())
                        }
                    }
        }

    }

    private companion object {

        val LOG: Logger = LoggerFactory.getLogger(GrpcFmuClient::class.java)

        private fun getReadRequest(instanceName: String, vr: List<ValueReference>): Service.ReadRequest {
            return Service.ReadRequest.newBuilder()
                    .setInstanceId(instanceName)
                    .addAllValueReferences(vr)
                    .build()
        }

    }

}
